Mathlib/Algebra/Star/NonUnitalSubalgebra.lean

```diff
@@ -0,0 +1,1224 @@
+
+theorem range_comp_le_range (f : A →⋆ₙₐ[R] B) (g : B →⋆ₙₐ[R] C) :
+  NonUnitalStarAlgHom.range (g.comp f) ≤ NonUnitalStarAlgHom.range g :=
+  SetLike.coe_mono (Set.range_comp_subset_range f g)
+#align non_unital_star_alg_hom.range_comp_le_range NonUnitalStarAlgHom.range_comp_le_range
+
+/-- Restrict the codomain of a non-unital star algebra homomorphism. -/
+def codRestrict (f : F) (S : NonUnitalStarSubalgebra R B) (hf : ∀ x, f x ∈ S) : A →⋆ₙₐ[R] S where
+  toNonUnitalAlgHom := NonUnitalAlgHom.codRestrict f S.toNonUnitalSubalgebra hf
+  map_star' := fun a => Subtype.ext <| map_star f a
+#align non_unital_star_alg_hom.cod_restrict NonUnitalStarAlgHom.codRestrict
+
+@[simp]
+theorem subtype_comp_codRestrict (f : F) (S : NonUnitalStarSubalgebra R B) (hf : ∀ x : A, f x ∈ S) :
+    (NonUnitalStarSubalgebraClass.subtype S).comp (NonUnitalStarAlgHom.codRestrict f S hf) = f :=
+  NonUnitalStarAlgHom.ext fun _ => rfl
+#align non_unital_star_alg_hom.subtype_comp_cod_restrict NonUnitalStarAlgHom.subtype_comp_codRestrict
+
+@[simp]
+theorem coe_codRestrict (f : F) (S : NonUnitalStarSubalgebra R B) (hf : ∀ x, f x ∈ S) (x : A) :
+    ↑(NonUnitalStarAlgHom.codRestrict f S hf x) = f x :=
+  rfl
+#align non_unital_star_alg_hom.coe_cod_restrict NonUnitalStarAlgHom.coe_codRestrict
+
+theorem injective_codRestrict (f : F) (S : NonUnitalStarSubalgebra R B) (hf : ∀ x : A, f x ∈ S) :
+    Function.Injective (NonUnitalStarAlgHom.codRestrict f S hf) ↔ Function.Injective f :=
+  ⟨fun H _x _y hxy => H <| Subtype.eq hxy, fun H _x _y hxy => H (congr_arg Subtype.val hxy : _)⟩
+#align non_unital_star_alg_hom.injective_cod_restrict NonUnitalStarAlgHom.injective_codRestrict
+
+/-- Restrict the codomain of a non-unital star algebra homomorphism `f` to `f.range`.
+
+This is the bundled version of `Set.rangeFactorization`. -/
+@[reducible]
+def rangeRestrict (f : F) : A →⋆ₙₐ[R] (NonUnitalStarAlgHom.range f : NonUnitalStarSubalgebra R B) :=
+  NonUnitalStarAlgHom.codRestrict f (NonUnitalStarAlgHom.range f)
+    (NonUnitalStarAlgHom.mem_range_self f)
+#align non_unital_star_alg_hom.range_restrict NonUnitalStarAlgHom.rangeRestrict
+
+/-- The equalizer of two non-unital star `R`-algebra homomorphisms -/
+def equalizer (ϕ ψ : F) : NonUnitalStarSubalgebra R A where
+  toNonUnitalSubalgebra := NonUnitalAlgHom.equalizer ϕ ψ
+  star_mem' := @fun x (hx : ϕ x = ψ x) => by simp [map_star, hx]
+#align non_unital_star_alg_hom.equalizer NonUnitalStarAlgHom.equalizer
+
+@[simp]
+theorem mem_equalizer (φ ψ : F) (x : A) :
+    x ∈ NonUnitalStarAlgHom.equalizer φ ψ ↔ φ x = ψ x :=
+  Iff.rfl
+#align non_unital_star_alg_hom.mem_equalizer NonUnitalStarAlgHom.mem_equalizer
+
+/-- The range of a morphism of non-unital star algebras is a fintype, if the domain is a fintype.
+
+Note that this instance can cause a diamond with `Subtype.fintype` if `B` is also a fintype. -/
+instance fintypeRange [Fintype A] [DecidableEq B] (φ : F) :
+    Fintype (NonUnitalAlgHom.range φ) :=
```

Comment:
How about:
```suggestion
instance finiteRange [Finite A] (φ : F) :
    Finite (NonUnitalAlgHom.range φ) :=
```
even though it diverges from `Set.fintypeRange`.

There would be no diamond issues!

Reply:
I've just deleted this entirely, as the domain will almost never be finite. If we ever want it, it's always possible to add as the proof is just `Set.Finite.finite_range φ`, but I expect this to come up basically never.